firmware_motion_board/src/protocol.c

```diff
@@ -0,0 +1,235 @@
+#include "protocol.h"
+#include "control_system.h"
+
+/* Helper functions to pack/unpack vars to/from byte arrays */
+void protocol_pack_int16(uint8_t *buffer, int16_t val)
+{
+    buffer[1] = val & 0xFF;
+    buffer[0] = (val >> 8) & 0xFF;
+}
+
+void protocol_pack_uint32(uint8_t *buffer, uint32_t val)
+{
+    for (int i = 3; i >= 0; i--)
+    {
+        buffer[i] = (val >> 8 * (3 - i)) & 0xFF;
+    }
+}
+
+void protocol_pack_float(uint8_t *buffer, float val)
+{
+    uint32_t as_integer = *((uint32_t *) & val); // access float on byte level
+    protocol_pack_uint32(buffer, as_integer);
+}
+
+float protocol_unpack_float(uint8_t *buffer)
+{
+    uint32_t val = 0;
+    for (int i = 0; i < 4; i++)
+    {
+        val |= ((uint32_t) buffer[i] << 8 * (3 - i));
+    }
+
+    float as_float = *((float *) &val);
+
+    return as_float;
+}
+
+uint16_t protocol_unpack_uint16(uint8_t *buffer)
+{
+    uint16_t val = 0;
+    val = (uint16_t) buffer[0] << 8;
+    val |= (uint16_t) buffer[1];
+
+    return val;
+}
+
+int16_t protocol_unpack_int16(uint8_t *buffer)
+{
+    uint16_t val = 0;
+    val = (uint16_t) buffer[0] << 8;
+    val |= (uint16_t) buffer[1];
+
+    return (int16_t) val;
+}
+
+/****************************************************/
+
+void can_send_quick(uint32_t id, uint8_t length, uint8_t *data)
+{
+    while (CAN4_TxFIFOIsFull(0));
+    // OR'ed with 0x80000000 so id becomes extended
+    CAN4_MessageTransmit((id | 0x80000000), length, data, 0, CAN_MSG_TX_DATA_FRAME);
+}
+
+void protocol_parse_msg(uint32_t id, uint8_t length, uint8_t *data)
+{
+    if (id != CAN_BASE_ID || length == 0)
+        return;
+
+    uint8_t command = data[0];
+
+    switch (command)
+    {
+    case CMD_READ_ENCODERS:
+    {
+        uint32_t enc_left = 0, enc_right = 0;
+        enc_left = -(int32_t)QEI3_PositionGet();
+        enc_right = QEI1_PositionGet();

```

Comment:
Ja bih da mi izbjegavamo akronime
```suggestion
        encoder_right = QEI1_PositionGet();

```

Reply:
Izmenjeno.